<a href="https://colab.research.google.com/github/Roohiashahzad/Roohia_INFO5731_Spring2020/blob/main/RoohiaShahzad_In_class_exercise_08.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The eighth in-class-exercise (20 points in total, 3/30/2021)**

The data for this exercise is from the dataset you created from assignment three. Please perform answer the following questions based on your data:

## (1) (10 points) Write a python program to extract the sentiment related terms from the corpus. You may use python package such as polyglot or external lexicon resources in the question. Rank the sentiment related terms by frequency.

In [99]:
pip install git+https://github.com/aboSamoor/polyglot@master

  Cloning https://github.com/aboSamoor/polyglot (to revision master) to /tmp/pip-req-build-gzi_sjaf
  Running command git clone -q https://github.com/aboSamoor/polyglot /tmp/pip-req-build-gzi_sjaf
  Created wheel for polyglot: filename=polyglot-16.7.4-py2.py3-none-any.whl size=70645 sha256=b566d062e1b75103361e259332c06a82a937cebdcd686b7c2c42cd3ddbff08ee
  Stored in directory: /tmp/pip-ephem-wheel-cache-xfwzpjyr/wheels/03/17/b5/807422952f3d35b08ba61250681db71b90bd8c885aeef2d946
Successfully built polyglot


In [100]:
# Write your code here
!pip install polygot
!pip install pyicu
!pip install pycld2



In [101]:
!pip install morfessor

In [102]:
from polyglot.downloader import downloader
downloader.download("embeddings2.en")
downloader.download("sentiment2.en")
downloader.download("pos2.en")


[polyglot_data] Downloading package embeddings2.en to
[polyglot_data]     /root/polyglot_data...
[polyglot_data]   Package embeddings2.en is already up-to-date!
[polyglot_data] Downloading package sentiment2.en to
[polyglot_data]     /root/polyglot_data...
[polyglot_data]   Package sentiment2.en is already up-to-date!
[polyglot_data] Downloading package pos2.en to /root/polyglot_data...
[polyglot_data]   Package pos2.en is already up-to-date!


True

In [103]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [104]:
import pandas as pd
df = pd.read_csv('/content/TwitterSentimentData.csv')
df = df.loc[:, ~df.columns.str.contains('^Unnamed')]


In [105]:
from nltk.corpus import stopwords
from textblob import TextBlob
stop = stopwords.words('english')

df['CleanText'] = df['CleanText'].apply(lambda x: " ".join(x.lower() for x in str(x).split() if x not in stop))
df['CleanText'] = df['CleanText'].str.replace('[^\w\s]','')

df['CleanText'] = df['CleanText'].apply(lambda x: " ".join(x for x in x.split() if x not in stop))
df['CleanText'] = df['CleanText'].apply(lambda x: str(TextBlob(x).correct()))


In [106]:
%%bash
polyglot download sentiment2.en 


[polyglot_data] Downloading package sentiment2.en to
[polyglot_data]     /root/polyglot_data...
[polyglot_data]   Package sentiment2.en is already up-to-date!


In [108]:
from polyglot.text import Text
sentimental_words = []
for line in df['CleanText']:
  text = Text(line)
  for w in text.words:
    #if w.polarity != 0:
    sentimental_words.append(w)
sentimental_words

['b',
 'today',
 'massachusetts',
 'penalized',
 'rest',
 'wuhancoronavirus',
 'vaccination',
 'timeline',
 'yet',
 'still',
 'de',
 'he',
 'x',
 'a',
 'http',
 'co',
 'lwulsrpm',
 'w',
 'b',
 'shall',
 'forever',
 'wuhancoronavirus',
 'http',
 'co',
 'ipcmob',
 'cow',
 'b',
 'truth',
 'could',
 'def',
 'bc',
 'x',
 'n',
 'he',
 'x',
 'had',
 'he',
 'x',
 'x',
 'he',
 'x',
 'x',
 'he',
 'of',
 'x',
 'he',
 'had',
 'a',
 'he',
 'x',
 'x',
 'he',
 'axe',
 'x',
 'e',
 'he',
 'a',
 'x',
 'c',
 'he',
 'axe',
 'a',
 'n',
 'laborunfall',
 'sarco',
 'sarco',
 'virus',
 'pandemie',
 'n',
 'he',
 'x',
 'cab',
 'he',
 'x',
 'and',
 'he',
 'a',
 'a',
 'drlimengyan',
 'n',
 'c',
 'a',
 'he',
 'x',
 'a',
 'http',
 'co',
 's',
 'q',
 'j',
 'b',
 'state',
 'department',
 'lead',
 'investigator',
 'say',
 'flumanchu',
 'may',
 'escaped',
 'lab',
 'n',
 'http',
 'co',
 'eotcqetzj',
 'n',
 'via',
 'he',
 'x',
 'a',
 'http',
 'co',
 'q',
 'uqffjt',
 'b',
 'tears',
 'fired',
 'n',
 'he',
 'x',
 'c',
 'x',


In [109]:
from collections import Counter
frequencies = Counter(sentimental_words)
rank_df = pd.DataFrame(list(frequencies.most_common()), columns=['Words', 'Frequencies'])
rank_df.index = list(range(1, len(frequencies.most_common())+1))
rank_df

Words  Frequencies
1       x          416
2      he          358
3       a          195
4       n          102
5       b           78
..    ...          ...
414  kind            1
415   ice            1
416   ing            1
417    up            1
418   nan            1

[418 rows x 2 columns]

## (2) (10 points) Compare the performance of the following tools in sentiment identification: TextBlob (https://textblob.readthedocs.io/en/dev/), VADER (https://github.com/cjhutto/vaderSentiment), TFIDF-based Support Vector Machine (SVM) (Split your data into training and testing data). Take your own annotation as the standard answers. 

Reference code: https://towardsdatascience.com/fine-grained-sentiment-analysis-in-python-part-1-2697bb111ed4

In [110]:
def Encoder(senti):
  sentiment_encoded = 0
  if senti == 'Positive':
    sentiment_encoded = 1
  elif senti == 'Negative':
    sentiment_encoded = 2
  return sentiment_encoded

In [111]:
actual_sentiment = []
for senti in df['Sentiment']:
  actual_sentiment.append(Encoder(senti))

In [112]:
from textblob import TextBlob
textblob_sentiments = []
for line in df['CleanText']:
  polarity = TextBlob(line).sentiment.polarity
  if polarity > 0:
    textblob_sentiments.append(1)
  elif polarity < 0:
    textblob_sentiments.append(2)
  elif polarity == 0.0:
    textblob_sentiments.append(0)
text_blob_df = pd.DataFrame(list(zip(df['CleanText'], actual_sentiment, textblob_sentiments)), columns = ['Tweets', 'Actual', 'Predicted'])
text_blob_df

Tweets  Actual  Predicted
0   b today massachusetts penalized rest wuhancoro...       1          0
1   b shall forever wuhancoronavirus http co ipcmo...       0          2
2   b truth could def bc x n he x had he x x he x ...       1          0
3   b state department lead investigator say fluma...       2          0
4   b tears fired n he x c x he sac x http co nkac...       0          1
..                                                ...     ...        ...
73  b truth could def bc x n he x had he x x he x ...       1          0
74  b truth could def bc x n he x had he x x he x ...       1          0
75  b globaltimesnews thank god virtual tour anywa...       1          1
76  b march pm eat real time mapping wuhancoronavi...       1          1
77                                                nan       0          0

[78 rows x 3 columns]

In [113]:
from sklearn.metrics import f1_score, accuracy_score
text_blob_accuracy = accuracy_score(text_blob_df['Actual'], text_blob_df['Predicted'])*100
text_blob_f1_score = f1_score(text_blob_df['Actual'], text_blob_df['Predicted'], average = 'macro')
print("Text Blob Accuracy is {0} and F1 Score is {1}".format(text_blob_accuracy, text_blob_f1_score))

Text Blob Accuracy is 46.15384615384615 and F1 Score is 0.4447290640394089


In [114]:
import nltk
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [115]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
vader = SentimentIntensityAnalyzer()
vader_sentiments = []
for line in df['CleanText']:
  polarity = vader.polarity_scores(line)['compound']
  if polarity > 0:
    vader_sentiments.append(1)
  elif polarity < 0:
    vader_sentiments.append(2)
  elif polarity == 0.0:
    vader_sentiments.append(0)
vader_df = pd.DataFrame(list(zip(df['CleanText'], actual_sentiment, vader_sentiments)), columns = ['Tweets', 'Actual', 'Predicted'])
vader_df

Tweets  Actual  Predicted
0   b today massachusetts penalized rest wuhancoro...       1          0
1   b shall forever wuhancoronavirus http co ipcmo...       0          0
2   b truth could def bc x n he x had he x x he x ...       1          1
3   b state department lead investigator say fluma...       2          0
4   b tears fired n he x c x he sac x http co nkac...       0          2
..                                                ...     ...        ...
73  b truth could def bc x n he x had he x x he x ...       1          1
74  b truth could def bc x n he x had he x x he x ...       1          1
75  b globaltimesnews thank god virtual tour anywa...       1          1
76  b march pm eat real time mapping wuhancoronavi...       1          0
77                                                nan       0          0

[78 rows x 3 columns]

In [116]:
from sklearn.metrics import f1_score, accuracy_score
vader_accuracy = accuracy_score(vader_df['Actual'], vader_df['Predicted'])*100
vader_f1_score = f1_score(vader_df['Actual'], vader_df['Predicted'], average = 'macro')
print("Vader Accuracy is {0} and F1 Score is {1}".format(vader_accuracy, vader_f1_score))

Vader Accuracy is 47.43589743589743 and F1 Score is 0.4851481069501091


In [123]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer

In [124]:
TFid_vectorizer = TfidfVectorizer(min_df = 5,
                             max_df = 0.8,
                             sublinear_tf = True,
                             use_idf = True)

In [140]:
import numpy as np
Train_X, Test_X, Train_Y, Test_Y = train_test_split(df['CleanText'],df['Sentiment'],test_size=0.33)
Encoder = LabelEncoder()
#df['CleanText'] = df['CleanText'].replace(np.nan, 'none', regex=True)
#df['Sentiment'] = df['Sentiment'].replace(np.nan, 'none', regex=True)
Train_Y = Encoder.fit_transform(Train_Y)
Test_Y = Encoder.fit_transform(Test_Y)
Train_X_vectors = TFid_vectorizer.fit_transform(Train_X)
Test_X_vectors = TFid_vectorizer.transform(Test_X)

In [141]:
from sklearn import svm
from sklearn.metrics import classification_report
svm_model = svm.SVC(kernel='linear')
svm_model.fit(Train_X_vectors, Train_Y)
predicted = svm_model.predict(Test_X_vectors)
report = classification_report(Test_Y, predicted, output_dict=True)
report

{'0': {'f1-score': 0.0, 'precision': 0.0, 'recall': 0.0, 'support': 6},
 '1': {'f1-score': 0.4666666666666667,
  'precision': 0.4666666666666667,
  'recall': 0.4666666666666667,
  'support': 15},
 '2': {'f1-score': 0.3076923076923077,
  'precision': 0.25,
  'recall': 0.4,
  'support': 5},
 'accuracy': 0.34615384615384615,
 'macro avg': {'f1-score': 0.25811965811965815,
  'precision': 0.2388888888888889,
  'recall': 0.2888888888888889,
  'support': 26},
 'weighted avg': {'f1-score': 0.32840236686390534,
  'precision': 0.3173076923076923,
  'recall': 0.34615384615384615,
  'support': 26}}


Text Blob accuracy = 46.15
Text Blob F1 score = 0.445


Vaber accuracy = 47.43
Vaber F1 score = 0.48

SVM accuracy = 34
SVM F1 score = 0

Vaber has higher F1 score than Text blob. SVM has the lowest F1 score. Vaber is better in comparison of all three options.
